Warn linux users about using --cask

apainintheneck opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is an alternate approach to #13925.

It does three things.
1. It raises an error when normal Linux users use the `--cask` switch.
2. It warns Linux developers (those with `HOMEBREW_DEVELOPER` set) after they use `--cask`.
3. It  more or less sets the `--formula` flag by default on Linux.
   - This is the addition to the `only_formula_or_cask` method which gets called inside of the `named_args` class when trying to determine formula and/or cask filtering.
   - It is used as the default value of the `:only` variable in a couple places as shown in the following example. https://github.com/Homebrew/brew/blob/7b00b2b1d70e045a5ad2fd952ed9b4b0dc2abd49/Library/Homebrew/cli/named_args.rb#L75-L80

apainintheneck commented:
Looks like this change broke a few tests related to making `--formula` the default behavior on Linux (third bullet point on the above list). It seems like we'd have to change a bunch of tests in `named_args` to be platform specific and I'm not sure that's really worth it.

apainintheneck commented on Library/Homebrew/test/spec_helper.rb:
```diff
@@ -180,6 +180,10 @@
     skip "Unzip is not installed." unless which("unzip")
   end
 
+  config.before(:each, :dev_on_linux) do
```
The advantage of this helper is that should we decide to remove support for casks on Linux completely, even for developers, we'd be able to just remove this and replace all instances of `:dev_on_linux` with `:needs_macos`.

MikeMcQuaid replied:
I was initially unsure about this but I really like this reasoning!

apainintheneck commented:
@SMillerDev I'd appreciate getting your feedback on this as well.

SMillerDev commented:
I don't know much about this code, I only know the issues people had and it seems to solve that. So lgtm

MikeMcQuaid commented on Library/Homebrew/cli/args.rb:
```diff
@@ -96,6 +96,7 @@ def context
       end
 
       def only_formula_or_cask
+        return :formula if OS.linux? && !Homebrew::EnvConfig.developer?
         return :formula if formula? && !cask?
         return :cask if cask? && !formula?
       end
```
@apainintheneck Ideally move to `extend/os` if you can be bothered, thanks ❤️

MikeMcQuaid commented on Library/Homebrew/cli/parser.rb:
```diff
@@ -341,14 +341,20 @@ def parse(argv = ARGV.freeze, ignore_invalid_options: false)
         @args
       end
 
+      def validate_options; end
```
Awesome!

apainintheneck commented:
> Should be good to go after final `extend/os` moves.
> 
> Ideally we'd have an rubocop for no (or, for now, no more `OS.linux?`/`OS.mac?` usage outside `extend/os`.

Oh, okay. I hadn't realized that. Yeah, I'll change it here and either work on adding the rubocop check or at the very least open an issue about that.

apainintheneck commented:
I was not able to reproduce these failing tests locally and they seem unrelated to what I changed (I just moved that snippet of code to `extend/os` like we talked about). Flaky tests perhaps? All of the failing tests are in the `#audit_revision_and_version_scheme` section of the `dev-cmd/audit_spec.rb` file though only about half the tests in that section are failing.

https://github.com/Homebrew/brew/blob/f35f9c03bef95fb1cf164b340392f7983933311f/Library/Homebrew/test/dev-cmd/audit_spec.rb#L856-L1123

MikeMcQuaid commented:
> Yeah, I'll change it here and either work on adding the rubocop check or at the very least open an issue about that.

Thanks 🙇🏻 



> Flaky tests perhaps?

Yeh, perhaps. Have rerun and we'll see 🤞🏻

MikeMcQuaid commented:
Works, thanks again @apainintheneck!
